source/dem/particle_wall_nonlinear_force.cc

```diff
@@ -309,5 +309,85 @@ ParticleWallNonLinearForce<dim>::calculate_nonlinear_contact_force_and_torque(
                          rolling_resistance_torque);
 }
 
+template <int dim>
+void
+ParticleWallNonLinearForce<dim>::calculate_IB_particle_wall_contact_force(
+  particle_wall_contact_info_struct<dim> &contact_info,
+  Tensor<1, 3> &                          normal_force,
+  Tensor<1, 3> &                          tangential_force,
+  Tensor<1, 3> &                          tangential_torque,
+  Tensor<1, 3> &                          rolling_resistance_torque,
+  IBParticle<dim> &                       particle,
+  const double &                          wall_youngs_modulus,
+  const double &                          wall_poisson_ratio,
+  const double &                          wall_restitution_coefficient,
+  const double &                          wall_friction_coefficient,
+  const double &                          wall_rolling_friction_coefficient,
+  const double &                          dt,
+  const double &                          mass,
+  const double &                          radius)
+{
+  auto particle_properties                        = particle.get_properties();
+  particle_properties[DEM::PropertiesIndex::mass] = mass;
+  particle_properties[DEM::PropertiesIndex::type] = 0;
+  particle_properties[DEM::PropertiesIndex::dp]   = 2 * radius;
+
+  // DEM::PropertiesIndex::type is the first (0) property of particles in the
+  // DEM solver. For the IB particles, the first property is ID. For force and
+  // torque calculations, we need pair-wise properties (such as effective
+  // Young's modulus, effective coefficient of restitution, etc.) We rewrite
+  // these pair-wise properties by using the ID of IB particles (using
+  // DEM::PropertiesIndex::type) and use them in force calculations.
+  const unsigned int particle_type =
+    particle_properties[DEM::PropertiesIndex::type];
+
+  this->effective_youngs_modulus[particle_type] =
+    (particle.youngs_modulus * wall_youngs_modulus) /
+    (wall_youngs_modulus *
+       (1 - particle.poisson_ratio * particle.poisson_ratio) +
+     particle.youngs_modulus * (1 - wall_poisson_ratio * wall_poisson_ratio) +
+     DBL_MIN);
+
+  this->effective_shear_modulus[particle_type] =
+    (particle.youngs_modulus * wall_youngs_modulus) /
+    ((2 * wall_youngs_modulus * (2 - particle.poisson_ratio) *
+      (1 + particle.poisson_ratio)) +
+     (2 * particle.youngs_modulus * (2 - wall_poisson_ratio) *
+      (1 + wall_poisson_ratio)) +
+     DBL_MIN);
+
+  this->effective_coefficient_of_restitution[particle_type] =
+    2 * particle.restitution_coefficient * wall_restitution_coefficient /
+    (particle.restitution_coefficient + wall_restitution_coefficient + DBL_MIN);
```

Comment:
are these harmonic average also or not really? You seem to be repeating this opeartion a lot. Seems worth it to put into a small inlined function.